### PR TITLE
Full re-write of spiral vase

### PR DESCRIPTION
**Pull request by andrewboktor**
Fixes:
https://github.com/prusa3d/PrusaSlicer/issues/2841
https://github.com/bambulab/BambuStudio/issues/2744
https://github.com/slic3r/Slic3r/issues/5160
https://github.com/slic3r/Slic3r/issues/4360

- Adds transition_out layer at the top to prevent sharp edge at the top of spiral vase. (cannot be turned on/off)
- Adds XY interpolation, configurable param for max distance for interpolation (absolute or % of nozzle_diameter, default 200%)
- Adds option to turn XY interpolation on/off

image
image
image
image
image
image

### CODE REVIEW

**Comment by andrewboktor** [comment]
> IMPORTANT: I really need help getting line_width in SpiralVase.cpp. As is right now, this will only work properly for line widths < 0.8mm or so.

Should we just add a new config "spiral_vase_max_xy_smoothing" that defaults to outer wall line width? What do y'all think about that idea?

**Comment by SoftFever** [comment]
> IMPORTANT: I really need help getting line_width in SpiralVase.cpp. As is right now, this will only work properly for line widths < 0.8mm or so.

Hi @andrewboktor 
In vase mode, we use outer_wall_line_width to define the shell thickness, which is set in the PrintRegionConfig. Currently, the SpiralVase class takes a PrintConfig parameter in its constructor. We could modify the class to accept an additional PrintRegionConfig parameter.

Then, the configuration can be passed as shown bellow(in GCode.cpp:1907)
```
    if (print.config().spiral_mode.value)
        m_spiral_vase = make_unique<SpiralVase>(print.config(),print.default_region_config());
```

**Comment by SoftFever** [comment]
> > IMPORTANT: I really need help getting line_width in SpiralVase.cpp. As is right now, this will only work properly for line widths < 0.8mm or so.
> 
> Should we just add a new config "spiral_vase_max_xy_smoothing" that defaults to outer wall line width? What do y'all think about that idea?

Adding this new parameter makes sense to me

File: src/libslic3r/GCode/SpiralVase.cpp [hunk 1]
```diff
@@ -70,28 +138,69 @@ std::string SpiralVase::process_layer(const std::string &gcode)
...
+                        line.set(reader, Z, z + factor * layer_height);
+                        if (smooth_spiral) {
+                            // Now we also need to try to interpolate X and Y
+                            SpiralPoint p(line.x(), line.y()); // Get current x/y coordinates
+                            current_layer->push_back(p);       // Store that point for later use on the next layer
+
+                            if (previous_layer != NULL) {
+                                bool        found    = false;
+                                float       dist     = 0;
+                                SpiralPoint nearestp = nearest_point_on_polygon(p, previous_layer, found, dist);
```
Comment (SoftFever) [inline src/libslic3r/GCode/SpiralVase.cpp:165]: Just thinking out loud here.
Not sure if we should add a normal check for the nearest point.
Theoretically such edge cases exists that, but I can't think of a real world model to test it.

image

**Comment by Noisyfox** [comment]
> > IMPORTANT: I really need help getting line_width in SpiralVase.cpp. As is right now, this will only work properly for line widths < 0.8mm or so.
> 
> Hi @andrewboktor In vase mode, we use outer_wall_line_width to define the shell thickness, which is set in the PrintRegionConfig. Currently, the SpiralVase class takes a PrintConfig parameter in its constructor. We could modify the class to accept an additional PrintRegionConfig parameter.
> 
> Then, the configuration can be passed as shown bellow(in GCode.cpp:1907)
> 
> ```
>     if (print.config().spiral_mode.value)
>         m_spiral_vase = make_unique<SpiralVase>(print.config(),print.default_region_config());
> ```

Ah this could be much simpler if no variable extrusion width is used then. And what if there are multiple objects that uses different line width?

**Comment by andrewboktor** [comment]
> Ah this could be much simpler if no variable extrusion width is used then. And what if there are multiple objects that uses different line width?
Can't we grab that setting "per object"? I don't know how to do that, but in theory it seems possible that we just grab the setting per object and that way each object will behave correction.

File: src/libslic3r/GCode/SpiralVase.cpp (hunk 1 above)
Comment (andrewboktor) [reply src/libslic3r/GCode/SpiralVase.cpp:165]: That's a good point. I think we need to come up with a real world example to justify writing the fix given how complex this is to deal with. Remember that smoothing occurs to the closest line, so you'd need this shape to exist on layer n and nothing similar on layer n-1 for the smoothing to go wrong.

**Comment by Noisyfox** [comment]
> > Ah this could be much simpler if no variable extrusion width is used then. And what if there are multiple objects that uses different line width?
> 
> Can't we grab that setting "per object"? I don't know how to do that, but in theory it seems possible that we just grab the setting per object and that way each object will behave correction.

The problem is this filter works per-layer, so you need a way to pass the corresponding object settings to it at each call. Fortunately vase mode does not support multiple objects with per-layer mode, so at least we don't need to deal with the problem that the gcode of a given layer comes from multiple objects which could have different line width.

File: src/libslic3r/GCode.cpp [hunk 2]
```diff
@@ -2729,12 +2729,13 @@ void GCode::process_layers(
...
    
     const auto spiral_mode = tbb::make_filter<LayerResult, LayerResult>(slic3r_tbb_filtermode::serial_in_order,
-        [&spiral_mode = *this->m_spiral_vase.get()](LayerResult in) -> LayerResult {
+        [&spiral_mode = *this->m_spiral_vase.get(), &layers_to_print](LayerResult in) -> LayerResult {
         	if (in.nop_layer_result)
                 return in;
                 
             spiral_mode.enable(in.spiral_vase_enable);
-            return { spiral_mode.process_layer(std::move(in.gcode)), in.layer_id, in.spiral_vase_enable, in.cooling_buffer_flush };
+            bool last_layer = in.layer_id==layers_to_print.size()-1;
```
Comment (SoftFever) [inline src/libslic3r/GCode.cpp:2737]: Pass the current printobject's region config here

**Comment by SoftFever** [comment]
To pass per-object wall width to the filter, we need to access the region config of the current PrintObject (as noted in the code review comments). In spiral mode, since there's typically only one region, taking the first region should suffice.

However, after giving it some thought, I'm inclined towards your second proposal to introduce a new parameter, spiral_vase_max_xy_smoothing. Just don't forget to make it accept either an absolute value or a percentage of the nozzle diameter, and we will be good.

**Comment by Noisyfox** [comment]
> nozzle diameter

nozzle diameter or line width?

**Comment by andrewboktor** [comment]
> > nozzle diameter
> 
> nozzle diameter or line width?

This should really be proportional to the object's outer_wall_line_width. So I guess there is not escaping getting and passing that setting properly. The new parameter doesn't solve any problems other than making things configurable so that if someone doesn't like what they see, then they can change it.

**Comment by SoftFever** [comment]
I meant nozzle diameter ;)  
Since it serves as a deviation cap, it doesn't matter much whether it's a percentage of the line width or the nozzle diameter.

File: src/libslic3r/GCode/SpiralVase.cpp (hunk 1 above)
Comment (SoftFever) [reply src/libslic3r/GCode/SpiralVase.cpp]: https://www.printables.com/model/432767-the-octahedroflake-a-higher-dimensional-analog-of-
This is probably the most closed model I can think of.

Comment (andrewboktor) [reply src/libslic3r/GCode/SpiralVase.cpp]: > https://www.printables.com/model/432767-the-octahedroflake-a-higher-dimensional-analog-of- This is probably the most closed model I can think of.

Yup, models like this are exactly why I made Smooth Spiral configurable on/off configurable. I would argue that models like this shouldn't even be printed in vase mode at all. They rely so heavily on one layer doing something then the next layer doing something completely different. That's not what vase mode is for.

**Comment by andrewboktor** [comment]
> I meant nozzle diameter(to save the hassle) ;) Since it serves as a deviation cap, it doesn't matter much whether it's a percentage of the line width or the nozzle diameter. But if you're keen on using the line width, by all means, go ahead. That would be ideal too

I thought about this long and hard. And I think we should make this configurable as a separate param, proportional to nozzle_diameter, get this PR in and get eyes on it and iterate from there. Otherwise we're probably gonna go down a (mostly unnecessary) rabbit hole.

**Comment by SoftFever** [comment]
> > I meant nozzle diameter(to save the hassle) ;) Since it serves as a deviation cap, it doesn't matter much whether it's a percentage of the line width or the nozzle diameter. But if you're keen on using the line width, by all means, go ahead. That would be ideal too
> 
> I thought about this long and hard. And I think we should make this configurable as a separate param, proportional to nozzle_diameter, get this PR in and get eyes on it and iterate from there. Otherwise we're probably gonna go down a (mostly unnecessary) rabbit hole.

Exactly my thought as well

**Comment by andrewboktor** [comment]
> Exactly my thought as well

Now the question is how do I get the right nozzle diameter. It seems that nozzle diameter is a floats, i.e. I have to find the right tool. Or should I just use nozzle_diameter[0]?

**Comment by andrewboktor** [comment]
> you can use EXTRUDER_CONFIG(nozzle_diameter); to get the corresponding nozzle diameter in process_layers function.

That worked fantastic! PR is now ready to go.

**Comment by Noisyfox** [comment]
image
Invalid gcode are generated.

**Comment by Noisyfox** [comment]
The issue was caused by the fact that the `dist_XY` is 0, as shown in the gcode that both lines have the exact same position
image

**Comment by Noisyfox** [comment]
image
This is the original gcode, which has a very small amount of the extrusion.

I suggest that if `dist_XY < EPSILON` then we do not modify the extrusion.

File: src/libslic3r/GCode/SpiralVase.cpp [hunk 3]
```diff
@@ -68,30 +135,69 @@ std::string SpiralVase::process_layer(const std::string &gcode)
                 return;
             } else {
                 float dist_XY = line.dist_XY(reader);
-                if (dist_XY > 0) {
```
Comment (Noisyfox) [inline src/libslic3r/GCode/SpiralVase.cpp:71]: See my other comments, this check needs to be kept, otherwise we could produce NaN.

**Review by Noisyfox — CHANGES_REQUESTED** [review]

**Comment by andrewboktor** [comment]
> The issue was caused by the fact that the `dist_XY` is 0, as shown in the gcode that both lines have the exact same position image

I can't repro this issue, but I understand what causes it. @Noisyfox Are you using just a primitive cone?

File: src/libslic3r/GCode/SpiralVase.hpp [hunk 4]
```diff
@@ -6,20 +6,32 @@
 
 namespace Slic3r {
 
+class SpiralPoint
```
[the remark this answers is not in the document]
Comment (andrewboktor) [reply src/libslic3r/GCode/SpiralVase.hpp:9]: Point class from Slic3r uses int coordinates, which won't work here.

File: src/libslic3r/GCode/SpiralVase.cpp (hunk 3 above)
Comment (andrewboktor) [reply src/libslic3r/GCode/SpiralVase.cpp:71]: On it, easy fix. We can't revert this bit to what it was before because of the zero length segments that appear in the gcode but we can definitely avoid doing the computation that explodes and just omit that segment altogether.

File: src/libslic3r/GCode/SpiralVase.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (andrewboktor) [reply src/libslic3r/GCode/SpiralVase.hpp]: This should be fixed now! I can't repro so @Noisyfox if you can give this another try or send me repro steps that would be great.

File: src/libslic3r/GCode/SpiralVase.cpp (hunk 3 above)
Comment (andrewboktor) [reply src/libslic3r/GCode/SpiralVase.cpp:71]: This should be fixed now! I can't repro so @Noisyfox if you can give this another try or send me repro steps that would be great.

**Comment by Noisyfox** [comment]
> > The issue was caused by the fact that the `dist_XY` is 0, as shown in the gcode that both lines have the exact same position image
> 
> I can't repro this issue, but I understand what causes it. @Noisyfox Are you using just a primitive cone?

I've uploaded a 3mf file, you should be able to reproducethe issue with it.

And ues it's a primitive cone.

**Comment by Noisyfox** [comment]
image
I can confirm it's solved, beautiful!

**Review by Noisyfox — APPROVED** [review]

**Comment by andrewboktor** [comment]
> 
> I've uploaded a 3mf file, you should be able to reproducethe issue with it.
> 
> And ues it's a primitive cone.

I just tried your 3mf on the latest version and it works great.

**Review by SoftFever — APPROVED** [review]
Looks good.
Cheers!

**Comment by m50434** [comment]
Hi @andrewboktor 
I have still some issues with the spriral vase mode.
If I only turn on the spiral vase without "Smoothing", my layers look like this:

grafik
grafik

Like you can see, there are some little things that doesnt look nice.

But if I now turn on smooth, its looks like this:

grafik
grafik

Not much better, on the contrary the edges have become much worse or uglier. This is what it looks like after printing.

Any ideas?

**Comment by andrewboktor** [comment]
@m50434 Can you share the 3mf and exact version (please retry on latest released version) used to expose those problems?
